lib/options.js

```diff
@@ -0,0 +1,151 @@
+    OptionsRequired: 'An options object is required.',
+    InputRequired: 'One of options.filename or options.fileContent is required.',
+    EncodingMissing: 'Internal Error: options.encoding is not set.',
+    EncodingFormat: ERROR_ENCODING_FORMAT + INFO_ENCODING_SUPPORTED,
+    EncodingNotSupported: (enc) => getUnsupportedEncodingString(enc),
+    IgnoredVisibilitiesMissing: 'Internal Error: options.ignoredVisibilities is not set.',
+    IgnoredVisibilitiesFormat: ERROR_VISIBILITIES_FORMAT + INFO_VISIBILITIES_SUPPORTED,
+    IgnoredVisibilitiesNotSupported: (arr) => `${getUnsupportedVisibilitiesString(arr)}`,
+});
+
+/** @type {BufferEncoding} */
+const DEFAULT_ENCODING = 'utf8';
+
+/** @type {SymbolVisibility[]} */
+const DEFAULT_IGNORED_VISIBILITIES = ['protected', 'private'];
+
+/** @returns {SvelteParserOptions} */
+function getDefaultOptions() {
+    return {
+        encoding: DEFAULT_ENCODING,
+        ignoredVisibilities: [...DEFAULT_IGNORED_VISIBILITIES],
+    };
+}
+
+function retrieveFileOptions(options) {
+    return {
+        structure: options.structure,
+        fileContent: options.fileContent,
+        filename: options.filename,
+        encoding: options.encoding,
+    };
+}
+
+/**
+ * Applies default values to options.
+ * @param {SvelteParserOptions} options
+ */
+function normalize(options) {
+    const defaults = getDefaultOptions();
+
+    Object.keys(defaults).forEach((optionKey) => {
+        /**
+         * If the key was not set by the user, apply default value.
+         * This is better than checking for falsy values because it catches
+         * use cases were a user tried to do something not intended with
+         * an option (e.g. putting a value of 'false' or an empty string)
+         */
+        if (!(optionKey in options)) {
+            options[optionKey] = defaults[optionKey];
+        }
+    });
+}
+
+/**
+ * @param {SvelteParserOptions} options
+ * @throws an error if any options are invalid
+ */
+function validate(options) {
+    if (!options) {
+        throw new Error(ErrorMessage.OptionsRequired);
+    }
+
+    normalize(options);
+
+    const hasFilename =
+        ('filename' in options) &&
+        isString(options.filename) &&
+        options.filename.length > 0;
+
+    // Don't check length for fileContent because it could be an empty file.
+    const hasFileContent =
+        ('fileContent' in options) &&
+        isString(options.fileContent);
+
+    if (!hasFilename && !hasFileContent) {
+        throw new Error(ErrorMessage.InputRequired);
+    }
+
+    if ('encoding' in options) {
+        if (!isString(options.encoding)) {
+            throw new Error(ErrorMessage.EncodingFormat);
+        }
+
+        if (!ENCODINGS.includes(options.encoding)) {
+            throw new Error(ErrorMessage.EncodingNotSupported(options.encoding));
+        }
+    } else {
+        // Sanity check. At this point, 'encoding' should be set.
+        throw new Error(ErrorMessage.EncodingMissing);
```

Comment:
Do we really always need the encoding? In case when source content are passed to the library by string value, the encoding is not needed.

Reply:
At this point, `normalize(options)` has already been called, which means a default encoding has been set on the `options` object. So this error should never being thrown.

Do you prefer If  `options.encoding` is set only if `filename` is present? The check could become :

```javascript
if ('encoding' in options) {
// ...
} else if (hasFilename) {
    // Sanity check: If filename is set, then 'encoding' should also be set.
    throw new Error(ErrorMessage.EncodingMissing);
}
```

Reply:
Ah, ok, lets keep as it is now